cedar-lean/Cedar/Thm/Core/LT.lean

```diff
@@ -433,16 +435,18 @@ theorem Values.lt_trans {vs₁ vs₂ vs₃: List Value}
   rename_i hd₁ tl₁ hd₂ tl₂ hd₃ tl₃
   rcases h₁ with h₁ | h₁ <;> rcases h₂ with h₂ | h₂
   case inl.inl =>
-    rcases (Value.lt_trans h₁ h₂) with h₃
+    have h₃ := Value.lt_trans h₁ h₂
     simp [LT.lt] at h₃ ; simp [h₃]
   case inl.inr =>
-    rcases h₂ with ⟨h₂, _⟩ ; subst h₂ ; simp [h₁]
+    have ⟨h₂, _⟩ := h₂
+    subst h₂ ; simp [h₁]
   case inr.inl =>
-    rcases h₁ with ⟨h₁, _⟩ ; subst h₁ ; simp [h₂]
+    have ⟨h₁, _⟩ := h₁
+    subst h₁ ; simp [h₂]
   case inr.inr =>
-    rcases h₁ with ⟨hl₁, h₁⟩ ; subst hl₁
-    rcases h₂ with ⟨hl₂, h₂⟩ ; subst hl₂
-    rcases (Values.lt_trans h₁ h₂) with h₃
+    have ⟨hl₁, h₃⟩ := h₁ ; subst hl₁
```

Comment:
I think we should use `rcases` in all cases that requires deconstructing the output, like this one.

Reply:
what's your rationale?

Reply:
I think `rcases` looks better in this case, and it makes the code style consistent.  

We clearly can't replace `rcases` with `have` everywhere, because it's a more powerful tactic.  So, I'd prefer a simple rule for when to use one versus the other:  use `rcases` when you need to deconstruct, and use `have` otherwise.

Using `rcases` also makes it possible / easier to apply a more advanced deconstruction pattern if that becomes necessary in the future as proofs change.

Reply:
I've thought about this more, and I propose a different simple rule for one versus the other: `rcases` for patterns involving `|`, and `have` in all other cases.

The reason I think this is sensible is that it reserves `rcases` for the times when we're actually splitting goals into separate _cases_, hence the name.  When we destructure an `and` with `⟨_, _⟩`, we're not casing on anything.  Lean splits our goal into 1 subcase called "intro", which is not helpful or elegant.

I'm skeptical that we will often be refactoring in a way that changes a term from a non-`|` pattern to a `|` pattern or vice versa.  And realistically, if we do need to refactor in that way, that will involve more invasive changes to the remainder of the proof (as we now have multiple cases to consider, or no longer have multiple cases to consider, respectively), so it's not a huge burden to change between `rcases` and `have` during that refactor.

Reply:
Personally I think `have` looks nicer than `rcases` for `⟨_, _⟩` expressions. The syntax looks sort of like a `let`, which makes it easier for me to remember.

Reply:
Thanks for pointing this out -- I think this is also the reason it feels more intuitive for me.  In most other languages, destructuring with `let` or an equivalent puts the bound variables on the left, not the right.

Reply:
I think `rcases` looks nicer because it looks like a `match` expression :)

Is it possible to use `have` when we're deconstructing an existential quantifier too?  And is it possible to use `have` for compound deconstruction patterns like `⟨a, ⟨b, c⟩, d⟩` ?  If not, then using `rcases` just for `|` won't work as a rule.

Reply:
Yes, `have` can deconstruct existentials, using `⟨a, h⟩` where `a` is the item that exists, and `h` is the new hypothesis that it satisfies.

Yes, `have` can do compound deconstructions like `⟨a, ⟨b, c⟩, d⟩`, and you can see an example of that in this PR.

Reply:
Example [here](https://github.com/cedar-policy/cedar-spec/pull/200/files#r1457850016)

Reply:
> To me, it's more intuitive to deconstruct with `match`es than with `let`s

For sum types, yes, but for product types, no.  I find this to be a weird way to destructure a value `tuple` of product type:
```
match tuple with
| (a, b) => ...
```
which is effectively what `rcases` with `⟨_, _⟩` patterns is doing.

Reply:
Ok, the rule works. 

But I still find having the matching pattern on the LHS really counterintuitive.  At least for all the languages I'm used to, the pattern-matching happens on the right.

I see why `have` is more appealing for binding a single name (a regular `let`).  But for deconstruction,`rcases` still looks better to me because it reads as `match`.

Curious:  is there any observable effect on proof performance (e.g., faster with `have`)?  

Or we just have to settle this based on the aesthetics?  :smile:

Reply:
CI took 2min 38sec to build the Lean for this PR, vs 2min 41sec to build the Lean in #199 (the previous change).  Could be noise, or could be that `have` is a small improvement vs `rcases`.  (I have to imagine that `have` is if anything faster, because since it is less powerful, it is doing less.  It's also not creating new subgoals like `rcases` is, even for trivial matches.)

Reply:
Yeah, I was hoping that `have` would be  much faster (because it's doing less), so that would be a good justification to go with `have` even for those who find it less intuitive :slightly_smiling_face:

I don't love it but I'll take the 3 seconds!

Reply:
Sounds like just a matter of preference then 🤷‍♀️ I usually reach for `let` over `match` because it avoids the extra indentation. (Which is not applicable in this case -- that's just why I like `let`)

Reply:
Agree, just a matter of preference.  And we have 2 vs 1 votes for pattern-matching on the left, so I'm disagreeing and committing on this one :)